rust/kernel/thread.rs

```diff
@@ -0,0 +1,404 @@
+            // there is no new thread to own the underlying object, we should
+            // let the current thread own it.
+            //
+            // SAFETY: We `from_pointer` back what we just `into_pointer`, and
+            // since the new thread is not created, so no one touches `data`.
+            unsafe {
+                A::from_pointer(data);
+            }
+
+            return Err(e);
+        }
+
+        result
+    }
+
+    /// Creates a new thread with a closure.
+    ///
+    /// Note that we have `'static` lifetime constraint for the closure, this
+    /// is because the new thread may outlive the caller of the function, we
+    /// need to make sure that any value that the closure captures is valid
+    /// until the thread exits, and the thread may run until the end of the
+    /// whole system, hence the `'static` lifetime.
+    ///
+    /// # Memory Cost
+    ///
+    /// The implementation of this function currently needs to allocate extra
+    /// dynamic memory (using [`Box`]), and do double boxing because
+    /// `Box<dyn FnOnce()>`. These "drawbacks" similarly exist in the
+    /// implementation of [`std::thread::spawn`]. Only use this when these
+    /// are acceptable, otherwise try [`Thread::try_new`].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use alloc::boxed::Box;
+    /// use kernel::thread::Thread;
+    ///
+    /// let mut a = 1;
+    /// let boxed_slice: Box<[i32]> = Box::try_new([1, 3, 4, 5])?; // Closures can capture DSTs.
+    ///
+    /// let t = Thread::try_new_closure(cstr!("rust-thread"), move || {
+    ///     let b = Box::try_new(42)?;
+    ///
+    ///     for _ in 0..10 {
+    ///         a = a + 1;
+    ///         println!("Hello Rust Thread {}", a + b.as_ref());
+    ///     }
+    ///
+    ///     println!("A Box<[i32]>: {:?}", boxed_slice);
+    ///     Ok(())
+    /// })?;
+    ///
+    /// t.wake_up();
+    /// ```
+    ///
+    /// # Context
+    ///
+    /// This function might sleep in `kthread_create_on_node` due to the memory
+    /// allocation and waiting for the completion, therefore do not call this
+    /// in atomic contexts (i.e. preemption-off contexts).
+    ///
+    /// [`std::thread::spawn`]: https://doc.rust-lang.org/std/thread/fn.spawn.html
+    pub fn try_new_closure<F>(name: CStr, f: F) -> KernelResult<Self>
+    where
+        F: FnOnce() -> KernelResult,
+        F: Send + 'static,
+    {
+        // Allocate closure here, because this function maybe returns before
+        // the bridged function (the function that uses the closure) get executed.
+        let boxed_fn: Box<dyn FnOnce() -> KernelResult + Send> = Box::try_new(f)?;
+
+        // Double boxing here because `dyn FnOnce` is a fat pointer, and we cannot
+        // `into_pointer` it to `*mut c_void`, see comments of `ClosureCallFunctor`
+        // for more.
+        let double_box = Box::try_new(boxed_fn)?;
+
+        Self::try_new::<_, ClosureCallFunctor>(name, double_box)
```

Comment:
Very nice that you now use `try_new` to avoid repeating code -- now this looks much cleaner!